tv_apps/smart-home/js/filter_manager.js

```diff
@@ -0,0 +1,142 @@
+    },
+    'DASHBOARD': {
+      name: 'dashboard',
+      iconName: 'dashboard'
+    },
+    'DEVICE': {
+      name: 'device',
+      iconName: 'device'
+    },
+    'APPLICATION': {
+      name: 'application',
+      iconName: 'application'
+    }
+  });
+
+  FilterManager.prototype = {
+    init: function(options) {
+      this._cardListElem = options.cardListElem;
+      this._cardScrollable = options.cardScrollable;
+      this._createCardNode = options.createCardNode;
+      this._cardManager = options.cardManager;
+
+      this._cardFilter = new CardFilter();
+      this._cardFilter.start(document.getElementById('filter-tab-group'));
+      this._cardFilter.filter = FilterManager.FILTERS.ALL.iconName;
+      this._filterChangedHandler = this.onFilterChanged.bind(this);
+      this._cardFilter.on('filterchanged', this._filterChangedHandler);
+
+      this._smartBubblesElem = document.getElementById('bubbles');
+
+      this._animationEndHandler = this.onAnimationEnd.bind(this);
+      this._smartBubblesElem.addEventListener('all-items-bubbled',
+        this._animationEndHandler);
+    },
+
+    uninit: function() {
+      this._cardFilter.off('filterchanged', this._filterChangedHandler);
+      this._smartBubblesElem.removeEventListener('all-items-bubbled',
+        this._animationEndHandler);
+    },
+
+    _filteredCardList: undefined,
+
+    _isFirstFrame: false,
+
+    _isFilterChanging: false,
+
+    _isBubbleSinking: function fm_isBubbleSinking() {
+      return (
+        this._cardListElem.getAttribute('smart-bubbles-direction') === 'down');
+    },
+
+    _performBubbleUp: function fm_performBubbleUp() {
+      if (this._isFirstFrame) {
+        this._isFirstFrame = false;
+        this._cardListElem.style.transition = 'none';
+        this._cardScrollable.resetScroll();
+        window.requestAnimationFrame(this._performBubbleUp.bind(this));
+      } else {
+        this._cardListElem.style.opacity = 1;
+        this._cardListElem.style.transition = undefined;
+        this._cardListElem.removeAttribute('smart-bubbles-direction');
+        this._smartBubblesElem.play(
+          document.querySelectorAll('#card-list > .card > .app-button'));
+      }
+    },
+
+    getFilterByIconName: function fm_getFilterByIconName(iconName) {
+      var keys = Object.keys(FilterManager.FILTERS);
+      var filter;
+      keys.some(function(key) {
+        var cursor = FilterManager.FILTERS[key];
+        if (iconName === cursor.iconName) {
+          filter = cursor;
+          return true;
+        }
+      });
+      return filter;
+    },
+
+    onAnimationEnd: function fm_onAnimationEnd() {
+      if (!this._isFilterChanging) {
+        return;
+      }
+
+      var that = this;
+      var filter = this.getFilterByIconName(this._cardFilter.filter);
+      var gotFilteredCardList = function(filteredList) {
+        filteredList.forEach(function(card) {
+          that._cardScrollable.addNode(that._createCardNode(card));
+        });
```

Comment:
Let's do 
1. pass `home` into `filterManager` on `init`.
2. Make `home._createCardNode` Public.

==

And if you get time to do further:
3. do `cardScrollable.clean()` and `cardScrollable.resetScroll()` in `home.createCardNode`.
4. remove `filterManager._cardScrollable`.